AVX512 Optimization for Gaussian Filter 2D

iqsxw opened the pull request:
ff_postscale_slice_c: 99
ff_postscale_slice_avx2: 5000
ff_postscale_slice_avx512: 5000

ff_horiz_slice_avx2(old): 109
ff_horiz_slice_avx2(new): 666.66
ff_horiz_slice_avx512: 1000

ff_verti_slice_c: 32.6
ff_verti_slice_avx2: 476.2
ff_verti_slice_avx512: 833.3

xhaihao commented on libavfilter/x86/vf_gblur.asm:
```diff
@@ -20,17 +20,371 @@
 cglobal horiz_slice, 4, 9, 9, ptr, width, height, steps, x, y, step, stride, remain
 %else
-cglobal horiz_slice, 4, 9, 9, ptr, width, height, steps, nu, bscale, x, y, step, stride, remain
+cglobal horiz_slice, 4, 9, 9 ptr, width, height, steps, nu, bscale, x, y, step, stride, remain
```
Is this change needed ? it seems you should not remove ',' here.

iqsxw replied:
No, it's not necessary at all. It was supposed to be deleted by me accidentally.

xuguangxin commented:
@iqsxw great! 
could you invite yanfei and other colleagues to review?
thanks

xuguangxin commented on libavfilter/x86/vf_gblur_init.c:
```diff
@@ -51,5 +55,17 @@ av_cold void ff_gblur_init_x86(GBlurContext *s)
+{
+    float *localbuf;
+
+    localbuf = (float *)av_malloc(16 * sizeof(float) * width);
```
we do not need malloc/free this again and again.
how about we move this malloc to higher level and pass it to this function. 
you can break this to two commits:
1. add temporary buffer to ff_postscale_slice_xxxx
2. implement ff_postscale_slice_avx512

iqsxw replied:
It seemed that we have to malloc again and again. The working of  horiz_slice was dispatched by worker queue of multi-threads. How and when each of horiz_slice that was called or done, was unpredictable, so each of horiz_slice needs a unique buffer to save their temp values to keep thread-safe.

xuguangxin replied:
maybe you can use jobnr

iqsxw replied:
Yeah. We might be able to change ThreadData to carry a local buffer variable and slice it use the values of jobnr.
But there are some questions. 
1. If we change ThreadData, which would affect other functions, there would be more things we need to update. When moving the local buffer to a higher level, the interface of functions would change. We might have to keep the function interface identical or we might be not able to pass the fate test.
2. Would it be a problem that I modify so much for one patch?

xuguangxin commented on libavfilter/x86/vf_gblur.asm:
```diff
@@ -20,17 +20,371 @@
+            ; Filter leftwards
+            dec xq
+            .loop_x_back:
+                ; Read 16x16 matrix from local buffer
```
not sure we can use some trick like this  or not
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=transpose&expand=3186,5961,5961

iqsxw replied:
It's already tested that using unpack and shuffle instructions like `_MM_TRANSPOSE4_PS ` is able to transpose 16x16 matrix, but it could decrease executing time 36.8% to use permute instruction to transpose matrix16x16. I think it's worthy to unroll for better performance.

HuiLi-Intel commented:
Great work!

iqsxw committed:
tests/checkasm/vf_gblur.c: add check_verti_slice() for unit test

Co-authored-by: Cheng Yanfei <yanfei.cheng@intel.com>
Co-authored-by: Jin Jun <jun.i.jin@intel.com>
Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

iqsxw committed:
tests/checkasm/vf_gblur.c: update check_horiz_slice for the new ff_horiz_slice_avx2/512

Co-authored-by: Cheng Yanfei <yanfei.cheng@intel.com>
Co-authored-by: Jin Jun <jun.i.jin@intel.com>
Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

iqsxw committed:
libavfilter/x86/vf_gblur: add ff_verti_slice_avx2/512()

The new vertical slice with AVX2/512 acceleration can significantly improve the performance of Gaussian Filter 2D.
Performance data (fps):
ff_verti_slice_c: 32.57
ff_verti_slice_avx2: 476.19
ff_verti_slice_avx512: 833.33

Co-authored-by: Cheng Yanfei <yanfei.cheng@intel.com>
Co-authored-by: Jin Jun <jun.i.jin@intel.com>
Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

iqsxw committed:
libavfilter/x86/vf_gblur: add localbuf and ff_horiz_slice_avx2/512()

We introduced a ff_horiz_slice_avx2/512() implemented on a new algorithm.
In a nutshell, the new algorithm does three things, gathering data from
8/16 rows, blurring data, and scattering data back to the image buffer.
Here we used a customized transpose 8x8/16x16 to avoid the huge overhead
brought by gather and scatter instructions, which is dependent on the
temporary buffer called localbuf added newly.

Performance data (fps):
ff_verti_slice_avx2(old): 109.89
ff_verti_slice_avx2(new): 666.67
ff_verti_slice_avx512: 1000

Co-authored-by: Cheng Yanfei <yanfei.cheng@intel.com>
Co-authored-by: Jin Jun <jun.i.jin@intel.com>
Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

iqsxw committed:
libavfilter/x86/vf_gblur: add ff_postscale_slice_avx512()

Co-authored-by: Cheng Yanfei <yanfei.cheng@intel.com>
Co-authored-by: Jin Jun <jun.i.jin@intel.com>
Signed-off-by: Wu Jianhua <jianhua.wu@intel.com>

xuguangxin commented:
@iqsxw please do not use merge, use "git pull -r "
other things are ok to me

yanfeich commented:
OK to me.

iqsxw committed:
Merge branch 'intel-media-ci:master' into master

uartie commented:
close. checks pass.  please reopen if you need more reviews and pre-check
